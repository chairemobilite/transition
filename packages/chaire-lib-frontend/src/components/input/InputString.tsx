/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React, { ChangeEvent, JSX } from 'react';

export type InputValue = {
    value: any;
    valid: boolean;
};

export type InputStringProps = {
    id: string;
    /**
     * @deprecated Use onValueUpdated instead. This method passes the value as
     * it is received by the event, callers had to check validity of the
     * received value. With the type and pattern props, the input itself can now
     * verify the validity, thus callers don't have to. If this prop is set,
     * onValueUpdated will be ignored.
     */
    onValueChange?: (e: any) => void;
    /**
     * Function called when the input value has been updated, passed with the
     * validity check of the method. Callers don't have to validate the value.
     * TODO: Make this value mandatory when onValueChange is removed
     */
    onValueUpdated?: (newValue: { value: any; valid: boolean }) => void;
    value?: string;
    maxLength?: number;
    disabled?: boolean;
    autocompleteChoices?: ({ label: string; value: string } | string)[];
    type?: 'text' | 'email' | 'number';
    pattern?: string;
};

type defaultInputType = {
    name: string;
    id: string;
    maxLength: number;
    value?: string;
    onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
    disabled?: boolean;
    pattern?: string;
};

const InputString: React.FunctionComponent<InputStringProps> = ({
    id,
    onValueChange = undefined,
    onValueUpdated = (_e) => {
        /* nothing to do */
    },
    value = undefined,
    maxLength = 255,
    disabled = false,
    autocompleteChoices = [],
    type = 'text',
    pattern = undefined
}: InputStringProps) => {
    const actualValue = value === undefined || value === null ? '' : value;
    const defaultAttributes: defaultInputType = {
        name: id,
        id: id,
        maxLength: maxLength,
        value: actualValue,
        onChange: onValueChange
            ? onValueChange
            : (evt: React.ChangeEvent<HTMLInputElement>) => {
                onValueUpdated({ value: evt.target.value, valid: evt.target.validity.valid });
            }
    };

    if (disabled === true) {
        defaultAttributes.disabled = true;
    }

    if (pattern) {
        defaultAttributes.pattern = pattern;
    }

    if (autocompleteChoices && autocompleteChoices.length > 0) {
        const dataListOptions: JSX.Element[] = [];
        autocompleteChoices.forEach((autocompleteChoice, key) => {
            dataListOptions.push(
                <option key={key}>
                    {typeof autocompleteChoice === 'string' ? autocompleteChoice : autocompleteChoice.label}
                </option>
            );
        });

        const dataListId = `datalist_for_${id}`;

        return (
            <React.Fragment>
                <input list={dataListId} {...defaultAttributes} />
                <datalist id={dataListId}>{dataListOptions}</datalist>
            </React.Fragment>
        );
    }

    return (
        <input
            type={type}
            autoComplete="none"
            className="apptr__form-input-string apptr__input apptr__input-string"
            {...defaultAttributes}
        />
    );
};

export default InputString;
