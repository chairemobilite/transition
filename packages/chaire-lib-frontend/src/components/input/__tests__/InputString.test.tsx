/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { create } from 'react-test-renderer';
import InputString from '../InputString';
import { render, fireEvent } from "@testing-library/react";
import { mount } from 'enzyme';

const mockOnChange = jest.fn();
const testId = "StringWidgetId";
const testLabel = "String label";

beforeEach(function () {
    mockOnChange.mockClear();
});

test('Default props', () => {
    const input = create(<InputString
        id = {testId}
        onValueUpdated = {mockOnChange}
    />)
        .toJSON();
    expect(input).toMatchSnapshot();
});

test('All props', () => {
    const input = create(<InputString
        id = {testId}
        onValueUpdated = {mockOnChange}
        value = "test"
        maxLength = {100}
        disabled = {false}
        autocompleteChoices = {[]}
        type = 'number'
        pattern = '[0-9]{1,2}'
    />)
        .toJSON();
    expect(input).toMatchSnapshot();
});

test('Disabled', () => {
    const input = create(<InputString
        id = {testId}
        onValueUpdated = {mockOnChange}
        disabled = {true}
    />)
        .toJSON();
    expect(input).toMatchSnapshot();
});

test('Autocomplete choices', () => {
    const input = create(<InputString
        id = {testId}
        onValueUpdated = {mockOnChange}
        autocompleteChoices = {[{label: "test", value: "test"},
            {label: "test1", value: "test1"}]}
    />)
        .toJSON();
    expect(input).toMatchSnapshot();
});

test('Default and initial empty value', () => {
    const value = "";
    const { getByLabelText } = render(
    <div>
        <InputString
            id = {testId}
            onValueUpdated = {mockOnChange}
            value = {value}
        />
        <label htmlFor={testId}>{testLabel}</label>
    </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    expect(input.value).toBe(value);
});

test('Call deprecated onChange', () => {
    const { getByLabelText } = render(
    <div>
        <InputString
            id = {testId}
            onValueChange = {mockOnChange}
        />
        <label htmlFor={testId}>{testLabel}</label>
    </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    const newText = "new text";
    fireEvent.change(input, {target: { value: newText}});
    expect(mockOnChange).toHaveBeenCalledTimes(1);
});

test('Call onValueUpdated', () => {
    const { getByLabelText } = render(
    <div>
        <InputString
            id = {testId}
            onValueUpdated = {mockOnChange}
        />
        <label htmlFor={testId}>{testLabel}</label>
    </div>);
    const input = getByLabelText(testLabel) as HTMLInputElement;
    const newText = "new text";
    fireEvent.change(input, {target: { value: newText}});
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith({value: newText, valid: true})
});

test('type check validity', () => {
    const originalIntValue = '5';
    const stringInput = mount(<InputString
        id = {testId}
        onValueUpdated = {mockOnChange}
        type = 'number'
        value = {originalIntValue}
    />);

    // Validate initial values
    const inputElements = stringInput.find({id: `${testId}`, type: 'number'});
    expect(inputElements).toBeTruthy();
    const inputElement = inputElements.last();
    expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe(originalIntValue);

    /* FIXME tahini: some tests fail here and behave differently than browser. Why?
    // Change the value manually to an invalid value
    const invalidValue = 'a';
    inputElement.getDOMNode<HTMLInputElement>().value = invalidValue;
    inputElement.simulate('change');
    expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe(invalidValue);
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith({value: invalidValue, valid: false});
    */

    // Go back to valid value
    const newIntValue = '10';
    inputElement.getDOMNode<HTMLInputElement>().value = newIntValue;
    inputElement.simulate('change');
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenLastCalledWith({value: newIntValue, valid: true});
    //expect(inputElement.getDOMNode<HTMLInputElement>().value).toBe(newIntValue);
    
})

