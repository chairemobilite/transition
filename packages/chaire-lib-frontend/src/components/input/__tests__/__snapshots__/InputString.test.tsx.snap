// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All props 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input-string apptr__input apptr__input-string"
  id="StringWidgetId"
  maxLength={100}
  name="StringWidgetId"
  onChange={[Function]}
  pattern="[0-9]{1,2}"
  type="number"
  value="test"
/>
`;

exports[`Autocomplete choices 1`] = `
[
  <input
    id="StringWidgetId"
    list="datalist_for_StringWidgetId"
    maxLength={255}
    name="StringWidgetId"
    onChange={[Function]}
    value=""
  />,
  <datalist
    id="datalist_for_StringWidgetId"
  >
    <option>
      test
    </option>
    <option>
      test1
    </option>
  </datalist>,
]
`;

exports[`Default props 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input-string apptr__input apptr__input-string"
  id="StringWidgetId"
  maxLength={255}
  name="StringWidgetId"
  onChange={[Function]}
  type="text"
  value=""
/>
`;

exports[`Disabled 1`] = `
<input
  autoComplete="none"
  className="apptr__form-input-string apptr__input apptr__input-string"
  disabled={true}
  id="StringWidgetId"
  maxLength={255}
  name="StringWidgetId"
  onChange={[Function]}
  type="text"
  value=""
/>
`;
